pinot-core/src/main/java/org/apache/pinot/core/transport/DirectOOMHandler.java

```diff
@@ -0,0 +1,94 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.transport;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.pinot.spi.exception.QueryCancelledException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+// Handling netty direct memory OOM. In this case there is a great chance that multiple channels are receiving
+// large data tables from servers concurrently. We want to close all channels to servers to proactively release
+// the direct memory, because the execution of netty threads can deadlock in allocating direct memory, in which case
+// no one will reach channelRead0.
+public class DirectOOMHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTableHandler.class);
+  private static final AtomicBoolean DIRECT_OOM_SHUTTING_DOWN = new AtomicBoolean(false);
+  private final QueryRouter _queryRouter;
+  private final ServerRoutingInstance _serverRoutingInstance;
+  private final ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> _serverToChannelMap;
+
+  public void setSilentShutDown() {
+    _silentShutDown = true;
+  }
+
+  private volatile boolean _silentShutDown = false;
+
+  public DirectOOMHandler(QueryRouter queryRouter, ServerRoutingInstance serverRoutingInstance,
+      ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> serverToChannelMap) {
+    _queryRouter = queryRouter;
+    _serverRoutingInstance = serverRoutingInstance;
+    _serverToChannelMap = serverToChannelMap;
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    if (_silentShutDown) {
+      return;
+    }
+    ctx.fireChannelInactive();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+    // catch direct memory oom here
+    if (cause instanceof OutOfMemoryError && cause.getMessage().contains("Direct buffer")) {
+      // only one thread can get here and do the shutdown
+      if (DIRECT_OOM_SHUTTING_DOWN.compareAndSet(false, true)) {
```

Comment:
The situation where this will evaluate to `false` is when a channel gets a DM OOM after another channel has OOMed so the channel shutdown process has already started? 

Should we have a log event for that? The channels are already being reset but I think we'd still want to know all the connections that got OOMed.

Reply:
> The situation where this will evaluate to false is when a channel gets a DM OOM after another channel has OOMed so the channel shutdown process has already started?

Yes

>Should we have a log event for that? The channels are already being reset but I think we'd still want to know all the connections that got OOMed.

Added for non-cleaner